core/src/main/java/org/elasticsearch/rest/action/get/RestHeadSourceAction.java

```diff
@@ -0,0 +1,71 @@
+/*
+ * Licensed to Elasticsearch under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.rest.action.get;
+
+import org.elasticsearch.action.get.GetRequest;
+import org.elasticsearch.action.get.GetResponse;
+import org.elasticsearch.client.Client;
+import org.elasticsearch.common.Strings;
+import org.elasticsearch.common.bytes.BytesArray;
+import org.elasticsearch.common.inject.Inject;
+import org.elasticsearch.common.settings.Settings;
+import org.elasticsearch.rest.*;
+import org.elasticsearch.rest.action.support.RestResponseListener;
+
+import static org.elasticsearch.rest.RestRequest.Method.HEAD;
+import static org.elasticsearch.rest.RestStatus.NOT_FOUND;
+import static org.elasticsearch.rest.RestStatus.OK;
+
+/**
+ * Identical to {@link RestHeadAction} except this handler checks if the document and its _source exist.
+ */
+public class RestHeadSourceAction extends BaseRestHandler {
```

Comment:
@martijnvg I will add some rest tests, or at least try :) Are they run as part of `gradle core:test`?
I created `RestHeadSourceAction` for a couple reasons: The GET actions were already split into 2 files: `RestGetAction` and `RestGetSourceAction` (IMO not ideal due to duplicated code, but I copied that pattern). Also `request.path().endsWith("_source")` is `false` for `HEAD test-index/test-type/1/_source/` (and also for `HEAD test-index/test-type/1/_source//` which may be a bug?) Is it possible to use the code in [PathTrie.java#L212](https://github.com/elastic/elasticsearch/blob/78d615f32054baeede280263ebc76bd11ea3951d/core/src/main/java/org/elasticsearch/common/path/PathTrie.java#L212) to break up the path by the separator and get the last part? If not I can write that logic myself and reuse `RestHeadAction` for `_source` as well. Would something like this work?
```
final boolean source = request.path().endsWith("_source"); // TODO figure out how to handle /_source/
client.get(getRequest, new RestResponseListener<GetResponse>(channel) {
    @Override
    public RestResponse buildResponse(GetResponse response) {
        if (!response.isExists()) {
            return new BytesRestResponse(NOT_FOUND, BytesRestResponse.TEXT_CONTENT_TYPE, BytesArray.EMPTY);
        } else if (source && response.isSourceEmpty()) { // doc exists, but source might not
            return new BytesRestResponse(NOT_FOUND, BytesRestResponse.TEXT_CONTENT_TYPE, BytesArray.EMPTY);
        } else {
            return new BytesRestResponse(OK, BytesRestResponse.TEXT_CONTENT_TYPE, BytesArray.EMPTY);
        }
    }
});
```
I also had another idea to do something clever in the `controller.registerHandler` call to distinguish the HEAD doc request vs HEAD doc _source request.